### ISSUE TITLE
False negatives with ignore: ["descendant"] in selector-no-type

### ISSUE THREAD

**Issue by jeddy3** [stylelint/stylelint#1946]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

Bug.

> Which rule, if any, is this issue related to?

`selector-no-type`

> What CSS is needed to reproduce this issue?

``` css
.class > a {} /* child */
.class + a {} /* next-sibling */
.class ~ a {} /* following-sibling */
```

> What stylelint configuration is needed to reproduce this issue?

``` json
{
  "rules": {
    "selector-no-type": [true, { "ignore": ["descendant"]}],
  }
}
```

> Which version of stylelint are you using?

`7.3.1`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

Via the demo

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

No

> What did you expect to happen?

The following warnings to be flagged:

``` shell
1:9 Unexpected type selector
2:9 Unexpected type selector
3:9 Unexpected type selector
```

> What actually happened (e.g. what warnings or errors you are getting)?

No warnings.

It looks like [this is bit of code](https://github.com/stylelint/stylelint/blob/master/lib/rules/selector-no-type/index.js#L82-L85) too permissive. Checking for _any_ type of combinators, rather than just the descendant one.

Ref: https://github.com/stylelint/stylelint/issues/1942

This is a breaking change. Post `8.0.0` this would be a _minor_ update, but prior to `8.0.0` it's a _major_.

**Comment by pixelass** [comment]
having the same issue. It is one of the most important features (IMHO) therefore I can't use it until it is fixed. It breaks the entire need for a linter.

**Comment by jeddy3** [comment]
> It is one of the most important features (IMHO) therefore I can't use it until it is fixed.

@pixelass Would you like to contribute this fix? We created a [developer userguide](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md) to help people get started in contributing to this open source project. All help is welcomed and appreciated.

**Comment by pixelass** [comment]
@jeddy3 sure, I'd love to contribute. sadly I'm very busy this month. Since I am using this at work I might be able to invest some of that time.

**Comment by pixelass** [comment]
I'm having a few problems running the tests. There are over 23000 tests which are all logged. 
Even if I test only this rule there are 90+ tests.

Wouldn't it make more sense to only logg errors? (not familiar with tape)
Maybe you there is a way to switch it off (at least while developing), otherwise it will be very hard and time consuming to fix issues.

**Comment by pixelass** [comment]
In my case the issue was caused by #1970 (still looking into this though)

**Comment by m-allanson** [comment]
I have a fix for this. Shall I rebase the v8 branch and PR into that?

**Comment by jeddy3** [comment]
> Shall I rebase the v8 branch and PR into that?

Yep, that sounds like a plan as `master` contains the Great Babel Removal.

**Comment by m-allanson** [comment]
Done in #2200
